File: apps/java/io/dustin/apps/common/model/ResponseWithScrollSetting.java
package io.dustin.apps.common.model;

import io.dustin.apps.common.exception.DataNotFoundException;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.util.Comparator;
import java.util.List;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ResponseWithScrollSetting {

    public static <T extends IdAble> CountByPagingInfo<T> getCountByPagingInfo(List<T> result, int realSize) {
        int querySize = realSize + 1;

        List<T> toClient;
        boolean isLast;
        Long nextId;
        if (result.size() <= realSize) {
            isLast = true;
            nextId = null;
            toClient = result;
        } else {
            isLast = false;
            toClient = result.subList(0, realSize);
            nextId = toClient.stream()
                    .sorted(Comparator.comparing(T::getId))
                    .findFirst().orElseThrow(() -> new DataNotFoundException("데이터에 문제가 있습니다.")).getId();
        }
        return new CountByPagingInfo(toClient, isLast, nextId);
    }
}

File: apps/src/main/kotlin/io/dustin/apps/common/model/ResponseWithScrollSetting.kt
import io.dustin.apps.common.exception.DataNotFoundException
import io.dustin.apps.common.model.CountByPagingInfo
import io.dustin.apps.common.model.IdAble
import java.util.Comparator

object ResponseWithScrollSetting {

    fun <T : IdAble> getCountByPagingInfo(result: List<T>, realSize: Int): CountByPagingInfo<T> {
        val querySize = realSize + 1

        val toClient: List<T>
        val isLast: Boolean
        var nextId: Int? = null

        if (result.size <= realSize) {
            isLast = true
            toClient = result
        } else {
            isLast = false
            toClient = result.subList(0, realSize)
            nextId = toClient.stream()
                .sorted(Comparator.comparing(T::id))
                .findFirst()
                .orElseThrow { DataNotFoundException("데이터에 문제가 있습니다.") }
                .id()


        return CountByPagingInfo(toClient, isLast, nextId)
    }
}
